src/dune_rules/gen_rules.ml

```diff
@@ -350,6 +350,13 @@ let automatic_sub_dirs_map =
     ; (Artifacts.Bin.bin_dir_basename, Bin)
     ]
 
+let automatic_subdirs path_components =
+  match List.last path_components with
```

Comment:
It's odd to first `explode` a directory into path components, thus constructing the whole list, and then immediately throw away all elements but the last one. How about we fuse the two steps?

Reply:
Used a slightly different approach in [e251d39](https://github.com/ocaml/dune/pull/7207/commits/e251d39b278b5de0f506b40920985f5464b7e828) that doesn't have these problems.

Reply:
> It's odd to first explode a directory into path components, thus constructing the whole list, and then immediately throw away all elements but the last one. How about we fuse the two steps?

Note that we aren't exploding the directory into path components in the rules. It's already exploded for us by the engine. So if the last component of the path is already provided for us, we should use it.

Reply:
@rgrinberg tried something like what you described in [04e678c](https://github.com/ocaml/dune/pull/7207/commits/04e678cf33c02caca8a87e3daf2a8e6f8b8922fe), though it's a bit more invasive. Which version do you prefer?

Reply:
> > It's odd to first explode a directory into path components, thus constructing the whole list, and then immediately throw away all elements but the last one. How about we fuse the two steps?
> 
> Note that we aren't exploding the directory into path components in the rules. It's already exploded for us by the engine. So if the last component of the path is already provided for us, we should use it.

@rgrinberg I'm not quite sure what you mean. In the version of the code which I was commenting on, we had an explicit `explode` followed by `List.last`. If we could get an exploded list from the caller, it would certainly be better but still, if we ever only look at the last component, the intermediate list of components is unnecessary (and maybe we need to change some API to avoid generating intermediate lists).

Reply:
There must have been an extra explode that I didn't notice. It's gone now I suppose.

>  if we ever only look at the last component, the intermediate list of components is unnecessary (and maybe we need to change some API to avoid generating intermediate lists).

We use the components elsewhere.

> (and maybe we need to change some API to avoid generating intermediate lists).

That would make sense. I don't see why the build engine has to pre-emptively split the path.